File: leetcode/leetcode200.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int numIslands(vector<vector<char>>& grid)
    {
        int m = grid.size(), n = grid[0].size(), ans = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j]=='1') {
                    ans++;
                    anya(grid, m, n, i, j);
                }
            }
        }
        return ans;
    }
    void anya(vector<vector<char>>& grid, int m, int n, int sr, int sc)
    {
        grid[sr][sc] = '0';
        if (sr + 1 < m && grid[sr + 1][sc]=='1') {
            anya(grid, m, n, sr + 1, sc);
        }
        if (sr - 1 >= 0 && grid[sr - 1][sc]=='1') {
            anya(grid, m, n, sr - 1, sc);
        }
        if (sc + 1 < n && grid[sr][sc + 1]=='1') {
            anya(grid, m, n, sr, sc + 1);
        }
        if (sc - 1 >= 0 && grid[sr][sc - 1]=='1') {
            anya(grid, m, n, sr, sc - 1);
        }
    }
};
int main(void)
{
    Solution s;
    vector<vector<int>> grid;
    vector<int> entry;
    entry.push_back(1);
    return 0;
}

File: leetcode/100-999/leetcode200.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        n, m = len(grid), len(grid[0])
        res = 0

        def chaewon(x, y):
            grid[x][y] = "0"
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < n and 0 <= ny < m and grid[nx][ny] == "1":
                    chaewon(nx, ny)

        for i in range(n):
            for j in range(m):
                if grid[i][j] == "1":
                    res += 1
                    chaewon(i, j)

        return res
